### DISCUSSION TITLE
A cool little ORM! :)

### DISCUSSION THREAD

**Discussion by ashercn97** [gleam-lang/gleam#3666] [Show and tell]
Hey guys!

I am new to Gleam, but I adore the language. I think it is super well-made and delightful to work with. 

I love Gleam's ability to compile to JS and Erlang, and I think it has a lot of potential for being a real language of the web. 

I've been doing web development for a while, and usually, I worked with lightweight backend "frameworks" (express in JS and flask in Python). Recently though, I discovered Phoenix in Elixir, and I was kind of hooked. It felt so much easier, and I could get much more done in a shorter amount of time.

Okay, so after all that, I noticed there was not really this kind of framework in Gleam. And there was not even a good ORM in Gleam!! So, I decided to make one.

Right now, it has a nice DSL for defining schemas, it has pretty decent code generation so you are writing less code, it has a query builder, more standard ORM things, and it all decodes into pretty little values so it is easy to use (i.e. you define a schema, generate types and decoders automatically, and then the queries to postgres decode into these types).

Is this of interest to anyone? Is it worth putting on GitHub?

**Comment by inoas** [comment]
While not an ORM for the dynamic query building part usually offered by ORM's you can look at https://hex.pm/packages/cake I guess.

I planned to build more ontop of it but so far have not found the time. I am curious to check out what you did :)

**Reply by ashercn97** [reply]
Ohhh thats so cool!

**Reply by ashercn97** [reply]
I was planning to make mine less query-builder and more like ORM-y. But this is a super cool project! Thanks for sharing!!! 

I will release mine, and ill let you know when I do.

**Reply by ashercn97** [reply]
Just came back to this. Would it be okay if forked the repo and like added features so it integrates with the schema/decoders/types from my project? Or would you rather I not. I am kinda new to this stuff so I am not sure the proper etiquette. Lmk!

**Reply by inoas** [reply]
I mean you can fork whatever you like :)

But what I would do is either integrate everyting in your lib and just have cake as a dependency or build a bridge library that sits between cake and your library.

Last but not least: If cake misses something that you really need and that requires changes to cake, do not hestitate to create an issue in the cake repo.

**Reply by inoas** [reply]
But whatever works for you :)

**Reply by ashercn97** [reply]
Thanks!

**Reply by inoas** [reply]
You can also get in touch on discord if you have questions.
Here are examples for adapters that still miss (I still need to create 4 companion libs for all 4 adapters to make it easy to use cake)

sqlite
https://discord.com/channels/768594524158427167/1265427201839599637/1289357475933454430
https://discord.com/channels/768594524158427167/1265427201839599637/1289357498477707345

postgres https://github.com/inoas/gleam-cake/blob/main/docs/demo-apps/helper/src/helper/postgres.gleam

**Reply by ashercn97** [reply]
Oh good to know! Thanks!

**Comment by ashercn97** [comment]
Its on GitHub! Check out [https://github.com/ashercn97/funsies](https://github.com/ashercn97/funsies)

**Comment by giacomocavalieri** [comment]
If you want to check it out there's also [squirrel](https://hex.pm/packages/squirrel), it goes the other way around by generating gleam code _from the sql queries._ I think it's really cool (but I'm biased as I'm the author 😆)

**Reply by ashercn97** [reply]
Yes, I saw squirrel! I thought it was literally so cool! I loved how seamless it was. Felt magical lol. 

Although, I was looking for a more Phoenix + Ecto-y tool, and thought that a more full-featured ORM would be neat too. 

I just made a CLI to do some more codegen with my project, and I was super inspired by Squirrel! Squirrel is such a cool tool (I can't say that enough!)

Knowing me, I will probably end up using Squirrel if I write something in Gleam because it is closer to how I've done things.

Sorry this was kind of rambly, but I just wanted to say Squirrel is a super cool project, but I guess mine has some different goals/approaches!
